Table summary and 'enhanced' translation functions

deepayan opened the pull request:
Two new functions: 

- `nhanesSummary()` gives per-variable summaries of an NHANES table using either the codebook or the dataset (or both), somewhat similar to `nhanesAttr()` but more detailed.

- `nhanesTranslateRaw()` (currently unexported) - similar to `nhanes(translated = TRUE)` but tries to be smarter in identifying common types of special missing-value codes.

Ref: <https://github.com/cjendres1/nhanes/issues/65>

cjendres1 reviewed (changes requested):
`nhanesSummary` I believe is too vague. Suggest we call it `nhanesTableSummary`  since it summarizes a table and its contents.

Regarding `nhanesTranslateRaw` , I have avoided doing any ad hoc manipulation of NHANES. I'm concerned that this function performs a cleansing operation that effectively alters the data. Want Robert and others to weigh in on whether or not we should be providing such functionality. If we go that route, I suggest we give such functions a distinctive prefix to make it clear that the function is not the expected default operation. E.g., cleansed_nhanesTranslate.

rgentlem commented:
Hi Chris and Deepayan,
   This is an interesting problem, and one I don't have a good idea on how
to fix.
The problem is the following - for a small number of variables, NHANES
codes them as numeric, but they then create one or more special
values, that are different from NA, as they want to encode different kinds
of missingness.  So, they make one type equal to 77777 and another type
equal to 9999 (for one variable I know about).
   This is a really bad idea - because it is hard to know about this fact -
you have to very carefully read the documentation...which few seem to do.
And not fixing the issue will give very bad results.  Personally I would
like to make the chance of that as small as I can. Which is why my personal
choice would be to modify these, possibly with a warning to the user.  But
I also understand that your philosophy is to change things as little as
possible.


 One way forward, that I think fits in with your paradigm for nhanesA is
that in the nhanesAttr output (or the new nhanesSummary - not clear to me
that we want two different functions to do such similar things) is to point
out that these variables are problematic - not sure just how to do that,
but perhaps an extra row (column?) that indicates whether a numeric
variable has unusual codes.  That would make it easier to deal with and we
could put an example in the vignette that shows how to find it, and one
suggestion for fixing it.

  We could then put the more aggressive table translation code into the
phonto package where it would be applied to the DB version by default. The
problem there is that the two will diverge - but in the long run that is a
bit inevitable since nhanesA doesn't have access to all the table
information, so it will not be able to do some things as they would be too
expensive (downloading everything for example).  For example, in solving
the issue above, we have in the container both the raw and the transformed
data - so anyone can easily compare.

   Part of the reason to adopt the Bioconductor File Caching capabilities
(I think that is in a separate pull request), is that using that software a
user would only ever download a file once, and after that it would use the
downloaded version.  Each time the file is accessed there is a check to see
if there is a newer version on the server, and if so the current version is
replaced.  I think in the long run this is a really good idea, as I think
that there is a lot of downloading the same file over and over that
happens.

  I am going to be out from Tuesday at noon, until the 27th, so could
schedule a meeting some morning of that week where we can discuss.

Robert


On Sun, Nov 12, 2023 at 9:24 AM Christopher Endres ***@***.***>
wrote:

> Assigned #66 <https://github.com/cjendres1/nhanes/pull/66> to @rgentlem
> <https://github.com/rgentlem>.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cjendres1/nhanes/pull/66#event-10934639422>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC7TWA6AMQCCAG3IMRLQFF3YEDL3LAVCNFSM6AAAAAA7G6JNE6VHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJQHEZTINRTHE2DEMQ>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>


-- 
Robert Gentleman
***@***.***

deepayan commented:
I think I prefer `nhanesTableSummary()`. I didn't want to replace `nhanesAttr()` because the results are now quite different, but even the `nhanesAttr()` name is not very informative. We could deprecate it at some point if that makes sense.

I have updated the PR accordingly.

For the "cleansing" version of `nhanesTranslate()`, I agree that it's not clear what the API should look like, but I also think that some version of it belongs in `nhanesA` because some of this is very basic --- effectively for numeric variables we are not doing any translation at all (even though the codebook flags special numeric codes), which is not consistent with what we are doing with categorical variables.

In the longer term, we may think about incorporating this into nhanesTranslate() with options like

`translate_numeric_codes = c("none", "minimal", "most")`

etc. For now, I think it's best to keep it internal to allow experimenting, so I have renamed it to `.nhanesTranslate_with_cleansing()`.

I haven't created a PR with the BiocFileCache changes yet (it needs a bit more work)

cjendres1 commented:
I'm open to changing the name of nhanesAttr and definitely prefer that it returns properties that don't require file download. It's the least used nhanesA function, so deprecating it or redefining it will have minimal impact on existing workflows.

cjendres1 commented:
Adding a "translate_numeric_codes" option seems like a promising and more natural solution, as opposed to creating a separate function. In general, I prefer more options if it means fewer functions.

deepayan commented:
Yes, I agree that a new argument like `translate_numeric_codes` to `nhanesTranslate()` is the optimal long term approach. 

We are not quite there yet, so should we go ahead and merge the PR with the internal function `.nhanesTranslate_with_cleansing()`?

cjendres1 reviewed (approved):
Will merge this, but I would like to establish a function naming convention. Any function that begins nhanes... (e.g. nhanesTranslate, nhanesTableVars, etc..) indicates that the function does not attempt by default to make any corrections or modifications to the data. Any function that does perform some modification should begin with something other than nhanes. E.g. cleansed_nhanesTranslate.

deepayan commented:
Sure, that makes sense. I will rename the internal function to something like `.translate_with_cleansing()` in the next PR.
